packages/react-dom/src/__tests__/validateDOMNesting-test.internal.js

```diff
@@ -9,176 +9,54 @@
 
 'use strict';
 
-var validateDOMNesting;
-
-// https://html.spec.whatwg.org/multipage/syntax.html#special
-var specialTags = [
-  'address',
-  'applet',
-  'area',
-  'article',
-  'aside',
-  'base',
-  'basefont',
-  'bgsound',
-  'blockquote',
-  'body',
-  'br',
-  'button',
-  'caption',
-  'center',
-  'col',
-  'colgroup',
-  'dd',
-  'details',
-  'dir',
-  'div',
-  'dl',
-  'dt',
-  'embed',
-  'fieldset',
-  'figcaption',
-  'figure',
-  'footer',
-  'form',
-  'frame',
-  'frameset',
-  'h1',
-  'h2',
-  'h3',
-  'h4',
-  'h5',
-  'h6',
-  'head',
-  'header',
-  'hgroup',
-  'hr',
-  'html',
-  'iframe',
-  'img',
-  'input',
-  'isindex',
-  'li',
-  'link',
-  'listing',
-  'main',
-  'marquee',
-  'menu',
-  'menuitem',
-  'meta',
-  'nav',
-  'noembed',
-  'noframes',
-  'noscript',
-  'object',
-  'ol',
-  'p',
-  'param',
-  'plaintext',
-  'pre',
-  'script',
-  'section',
-  'select',
-  'source',
-  'style',
-  'summary',
-  'table',
-  'tbody',
-  'td',
-  'template',
-  'textarea',
-  'tfoot',
-  'th',
-  'thead',
-  'title',
-  'tr',
-  'track',
-  'ul',
-  'wbr',
-  'xmp',
-];
-
-// https://html.spec.whatwg.org/multipage/syntax.html#formatting
-var formattingTags = [
-  'a',
-  'b',
-  'big',
-  'code',
-  'em',
-  'font',
-  'i',
-  'nobr',
-  's',
-  'small',
-  'strike',
-  'strong',
-  'tt',
-  'u',
-];
-
-function isTagStackValid(stack) {
-  var ancestorInfo = null;
-  for (var i = 0; i < stack.length; i++) {
-    if (!validateDOMNesting.isTagValidInContext(stack[i], ancestorInfo)) {
-      return false;
-    }
-    ancestorInfo = validateDOMNesting.updatedAncestorInfo(
-      ancestorInfo,
-      stack[i],
-      null,
-    );
+var React = require('react');
+var ReactDOM = require('react-dom');
+
+function expectInvalidNestingWarning(shouldWarn, tags) {
+  let element = null;
+  const container = document.createElement(tags.splice(0, 1));
```

Comment:
I think it's a bit confusing that this function can mutate its argument. Let's copy `tags` at the beginning of this function?